packages/jest-transform/src/ScriptTransformer.ts

```diff
@@ -209,21 +213,28 @@ export default class ScriptTransformer {
             // files outside `cwd` will not be instrumented
             cwd: this._config.rootDir,
             exclude: [],
+            // Needed for correct coverage as soon as we start storing a source map of the instrumented code
+            inputSourceMap: input.map,
             useInlineSourceMaps: false,
           },
         ],
       ],
+      /**
+       * Necessary to be able to map back to original source from the instrumented code
+       * the inline map is needed for debugging functionality
+       * and exposing it as a separate file is needed for e.g. mapping stack traces
+       * convenient to use 'both' here and avoid extracting the source map
+       *
+       * Prior behavior of emitting no map when we can't map back to original source is preserved
+       */
+      sourceMaps: canMapToInput ? 'both' : false,
     });
 
-    if (result) {
-      const {code} = result;
-
-      if (code) {
-        return code;
-      }
+    if (result && result.code) {
+      return {code: result.code, map: result.map};
```

Comment:
```suggestion
      return result;
```

Reply:
Typescript compiler doesn't seem to understand that `result.code` being truthy means `result` has a `code` property which is not `undefined`/`null`.

We can declare that we know what we're doing: `return result as TransformResult;`, which seems a little less awkward.